documentation/private-link-service/private-link-redirect-map.md

```diff
@@ -0,0 +1,261 @@
+
+## 4. Resource modeling, requirement, and SLA
+
+### 4.1 Scaling requirement
+
+For scaling requirement, please refer to [SONIC DASH HLD](https://github.com/sonic-net/DASH/blob/main/documentation/general/dash-sonic-hld.md#14-scaling-requirements).
+
+### 4.2 Reliability requirements
+
+The flows created when redirect map will be replicated among multiple DPUs, following the SmartSwitch HA design.
+
+For more information, please refer to [SmartSwitch HA design doc](https://github.com/sonic-net/SONiC/blob/master/doc/smart-switch/high-availability/smart-switch-ha-hld.md).
+
+## 5. SAI API design
+
+### 5.1. DASH Outbound Port Map
+
+A new object `outbound-port-map` is added to manage port mappings in the outbound direction, which includes the following attributes:
+
+| Attribute name | Type | Description |
+| --- | --- | --- |
+| SAI_OUTBOUND_PORT_MAP_ATTR_COUNTER_ID | sai_object_id_t | A counter that tracks the number of packets and bytes associated with the matching outbound-port-map. |
+
+### 5.2. DASH Outbound Port Map Port Range
+
+A new object key entry `outbound-port-map-port-range` is designed to define port range translation in the outbound direction. It is particularly useful for scenarios involving Private Link (PL) or Private Link Network Security Group (PL NSG) redirect maps. The key entry for this object specifies how port ranges should be translated when traffic is directed outbound.
+
+The entry is defined like below:
+
+```c
+typedef struct _sai_outbound_port_map_port_range_entry_t
+{
+    /**
+     * @brief Switch ID
+     *
+     * @objects SAI_OBJECT_TYPE_SWITCH
+     */
+    sai_object_id_t switch_id;
+
+    /**
+     * @brief Exact matched key outbound_port_map_id
+     *
+     * @objects SAI_OBJECT_TYPE_OUTBOUND_PORT_MAP
+     */
+    sai_object_id_t outbound_port_map_id;
+
+    /**
+     * @brief Range matched key dst_port_range
+     */
+    sai_u32_range_t dst_port_range;
+
+    /**
+     * @brief Rule priority in table
+     */
+    sai_uint32_t priority;
+
+} sai_outbound_port_map_port_range_entry_t;
+```
+
+The member `dst_port_range` designates the match field of port range, targeting packet TCP/UDP destination port. The member `priority` is a selection hint if more than one entry is matched, and the higher number takes priority. The entry includes the following attributes:
+
+| Attribute name | Type | Description |
+| --- | --- | --- |
+| SAI_OUTBOUND_PORT_MAP_PORT_RANGE_ENTRY_ATTR_ACTION | sai_outbound_port_map_port_range_entry_action_t | Action `skip_mapping` or `map_to_private_link_service`. |
+| SAI_OUTBOUND_PORT_MAP_PORT_RANGE_ENTRY_ATTR_BACKEND_IP | sai_ip_address_t | Back end IP, it overrides underlay destination IP |
+| SAI_OUTBOUND_PORT_MAP_PORT_RANGE_ENTRY_ATTR_MATCH_PORT_BASE | sai_uint16_t | Match port base, aka begin port of port range |
+| SAI_OUTBOUND_PORT_MAP_PORT_RANGE_ENTRY_ATTR_BACKEND_PORT_BASE | sai_uint16_t | Back end port base, aka begin port of translated port range |
```

Comment:
What is 'backend' referring to here? Could we use 'underlay' instead of 'backend' and specify 'DIP' or 'SIP' in the attribute name so that it's consistent across different DASH objects?

Reply:
Here 'backend' means back-end vm instance, locating after private link service VIP.

Reply:
hi @mukeshmv , I think calling it "Underlay DIP" will be somewhat confusing, because this IP is being used in both underlay and overlay. In the private link use case, this is the IP of the private link service backend, so calling it backend IP is closer to the use case.

If you think this is confusing or not flexible enough, maybe splitting it into 2 attributes UNDERLAY_IP and OVERLAY_IP will be better. Please let us know which way you prefer.

Reply:
hi @r12f, I think you mixed up my comment with Lawrence's comment. I am fine with the name BACKEND_IP.

My comment was about the description for BACKEND_IP. 
Why does it say it's also used as underlay IP? The underlay / outer IP comes from CA-PA mapping right? 
Is this Backend IP also used to overwrite the underlay IP in the case of Private link service? 
That does not make sense if this is the backend VM IP.

Reply:
> hi @mukeshmv , I think calling it "Underlay DIP" will be somewhat confusing, because this IP is being used in both underlay and overlay. In the private link use case, this is the IP of the private link service backend, so calling it backend IP is closer to the use case.
> 
> If you think this is confusing or not flexible enough, maybe splitting it into 2 attributes UNDERLAY_IP and OVERLAY_IP will be better. Please let us know which way you prefer.

didn't realize this was used in both overlay and underlay, agree w/ Mukesh about clarifying the description and keeping the existing name then.

Reply:
> Why does it say it's also used as underlay IP? The underlay / outer IP comes from CA-PA mapping right?
> Is this Backend IP also used to overwrite the underlay IP in the case of Private link service?

In this case, the backend IP is used in both overlay and underlay. The one in the CA-PA mapping will be override by the one in the port map entry.

@jimmyzhai , I think this description "Back end IP, it overrides underlay destination IP" is not complete, so it will be better to clarify it.

Reply:
Updated the description to "The IP of private link service backend, used in both overlay and underlay"